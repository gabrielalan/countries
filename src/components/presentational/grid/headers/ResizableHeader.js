import React, { useRef, useState } from 'react';

const dragGhostInvisibleImage = new Image();
dragGhostInvisibleImage.src = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+ip1sAAAAASUVORK5CYII=';

const getStyleValue = (element, property) => window
  .getComputedStyle(element)
  .getPropertyValue(property);

/**
 * This component represents the resizable headers
 * Once the user finishes to resize the header via DnD
 * it calls the callback onResize, to let the grid container
 * know the size for the column has changed.
 * 
 * It's really agnostic regarding the children or grid data.
 */
export default function ResizableHeader({ children, onResize, initialWidth }) {
  const div = useRef(undefined);
  const [width, setWidth] = useState(initialWidth);
  const [initials, setInitials] = useState({ width: initialWidth });

  const setResizedWidth = x => setWidth(initials.width + (x - initials.x));

  /**
   * Firefox has a known bug that makes screenX/clientX/pageX always 0
   * So the resize won't work for Firefox. The solution would be to use
   * onDragOver event, but that brings a lot of side-effects and is not a 
   * very performatic workaround.
   */
  const onDrag = e => e.screenX > 0 && setResizedWidth(e.screenX);

  const onDragStart = e => {
    e.dataTransfer.setData('text', ''); // required in some browsers
    e.dataTransfer.setDragImage(dragGhostInvisibleImage, 0, 0);
    
    const padLeft = getStyleValue(div.current, 'padding-left');
    const padRight = getStyleValue(div.current, 'padding-right');

    setInitials({
      x: e.screenX,
      width: div.current.getBoundingClientRect().width - (parseInt(padLeft) + parseInt(padRight))
    });
  };

  return <div ref={div} className="column" style={{ width }}>
    {children}
    <span
      onDragStart={onDragStart}
      onDragEnd={() => onResize(width)}
      onDrag={onDrag}
      className="resizer"
      draggable />
  </div>;
}