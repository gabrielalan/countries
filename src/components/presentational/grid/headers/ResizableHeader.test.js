import React from 'react';
import renderer from 'react-test-renderer';
import ResizableHeader from './ResizableHeader';

const onResizeSpy = jest.fn();

const dataTransfer = {
  setData() {},
  setDragImage() {}
};

const getComputedStyle = () => ({
  getPropertyValue: () => '0'
});

const createNodeMocker = width => (element) => {
  if (element.type === 'div') {
    return {
      getBoundingClientRect: () => ({ width })
    };
  }

  return null;
};

describe('ResizableHeader', () => {
  beforeAll(() => (global.getComputedStyle = getComputedStyle));

  beforeEach(() => onResizeSpy.mockClear());

  it('should set new width after drag', () => {
    let tree;

    renderer.act(() => {
      tree = renderer.create(
        <ResizableHeader onResize={onResizeSpy} initialWidth={100}>text</ResizableHeader>,
        { createNodeMock: createNodeMocker(100) }
      );
    });
    
    const span = tree.root.findByType('span');

    expect(tree.toJSON()).toMatchSnapshot();

    renderer.act(() => {
      span.props.onDragStart({
        dataTransfer,
        pageX: 50
      });
    });
      
    renderer.act(() => {
      span.props.onDrag({ pageX: 100 });
    });

    renderer.act(() => {
      span.props.onDragEnd({ pageX: 100 });
    });

    expect(onResizeSpy).toBeCalledWith(150);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
