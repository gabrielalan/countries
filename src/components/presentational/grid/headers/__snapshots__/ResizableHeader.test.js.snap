// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizableHeader should set new width after drag 1`] = `
<div
  className="grid__row__column"
  style={
    Object {
      "width": 100,
    }
  }
>
  text
  <span
    className="grid__row__column__resizer"
    draggable={true}
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  />
</div>
`;

exports[`ResizableHeader should set new width after drag 2`] = `
<div
  className="grid__row__column"
  style={
    Object {
      "width": 150,
    }
  }
>
  text
  <span
    className="grid__row__column__resizer"
    draggable={true}
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  />
</div>
`;
